### PR TITLE
PS Vita fixes and gamepad mapping improvements

### PR DESCRIPTION
**Pull request by Northfear**
Bunch of fixes related to PS Vita
- Disabled rear touchpad
- Added gamepad mapping for cursor speedup
- Fixed fullscreen/windowed toggle in display options

### CODE REVIEW

File: src/engine/localevent.cpp [hunk 1]
```diff
@@ -1571,16 +1577,23 @@ void LocalEvent::HandleControllerButtonEvent( const SDL_ControllerButtonEvent &
...
+        }
+        else if ( button.button == SDL_CONTROLLER_BUTTON_DPAD_DOWN ) {
             key_value = fheroes2::Key::KEY_SPACE;
         }
-        else if ( button.button == SDL_CONTROLLER_BUTTON_LEFTSHOULDER ) {
+        else if ( button.button == SDL_CONTROLLER_BUTTON_DPAD_LEFT ) {
             key_value = fheroes2::Key::KEY_H;
         }
-        else if ( button.button == SDL_CONTROLLER_BUTTON_RIGHTSHOULDER ) {
+        else if ( button.button == SDL_CONTROLLER_BUTTON_DPAD_RIGHT ) {
```
Comment (ihhub) [inline src/engine/localevent.cpp:1596]: Are these changes only Vita specific? Would they affect Switch?

File: src/engine/screen.cpp [hunk 2]
```diff
@@ -733,6 +714,38 @@ namespace
         {
             updatePalette( StandardPaletteIndexes() );
         }
+
+        void _calculateScreenScaling( int32_t width_, int32_t height_, bool isFullScreen )
```
Comment (ihhub) [inline src/engine/screen.cpp:718]: Please make all input arguments as `const`.

File: src/engine/screen.cpp [hunk 3]
```diff
@@ -733,6 +714,38 @@ namespace
...
+        {
+            _destRect.x = 0;
+            _destRect.y = 0;
+            _destRect.width = width_;
+            _destRect.height = height_;
+
+            if ( width_ != VITA_FULLSCREEN_WIDTH || height_ != VITA_FULLSCREEN_HEIGHT ) {
```
Comment (ihhub) [inline src/engine/screen.cpp:725]: Let's do the opposite logic to remove extra text indentation of the code:
```cpp
 if ( width_ == VITA_FULLSCREEN_WIDTH && height_ == VITA_FULLSCREEN_HEIGHT ) {
    // Nothing to do more.
    return;
}

if ( !isFullScreen ) {
    // Center game area.
    _destRect.x = ( VITA_FULLSCREEN_WIDTH - width_ ) / 2;
    _destRect.y = ( VITA_FULLSCREEN_HEIGHT - height_ ) / 2;
    return;
}

...
the rest of the code
```

File: src/engine/screen.cpp [hunk 4]
```diff
@@ -733,6 +714,38 @@ namespace
...
+                if ( isFullScreen ) {
+                    vita2d_texture_set_filters( _texBuffer, isNearestScaling() ? SCE_GXM_TEXTURE_FILTER_POINT : SCE_GXM_TEXTURE_FILTER_LINEAR,
```
Comment (ihhub) [inline src/engine/screen.cpp:727]: Can we create a separate variable for:
```cpp
isNearestScaling() ? SCE_GXM_TEXTURE_FILTER_POINT : SCE_GXM_TEXTURE_FILTER_LINEAR
```
and use it within **vita2d_texture_set_filters** call to reduce code duplication and possible bugs?

File: src/engine/screen.cpp [hunk 5]
```diff
@@ -550,6 +550,14 @@ namespace
 
         RenderEngine & operator=( const RenderEngine & ) = delete;
 
+        void toggleFullScreen() override
```
Comment (ihhub) [inline src/engine/screen.cpp:553]: In this case should we modify `bool System::isHandheldDevice()` to set fullscreen by default for Vita and Switch?

File: src/engine/localevent.cpp [hunk 6]
```diff
@@ -1615,8 +1631,16 @@ void LocalEvent::HandleControllerButtonEvent( const SDL_ControllerButtonEvent &
         else if ( button.button == SWITCH_BUTTON_PLUS ) {
             key_value = fheroes2::Key::KEY_C;
         }
+        else {
+            key_value = fheroes2::Key::NONE;
+        }
 #endif
     }
+    else {
+        if ( button.button == SDL_CONTROLLER_BUTTON_RIGHTSHOULDER ) {
```
Comment (ihhub) [inline src/engine/localevent.cpp:1640]: Please combine these else and if statements.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Northfear , I added few comments here. Would you mind to take a look when you have time?

File: src/engine/localevent.cpp (hunk 1 above)
Comment (Northfear) [reply src/engine/localevent.cpp:1596]: Yeah, should affect anything with a gamepad. But those changes are pretty good IMO. DPad buttons were free anyway and "cursor speedup" is really useful for thumbstick cursor control.

File: src/engine/screen.cpp (hunk 5 above)
Comment (Northfear) [reply src/engine/screen.cpp:553]: Hard to say. I like the "sharp" look of unscaled image. Even if it's not fullscreen.
Most people prefer fullscreen image tho, so yeah. Defaulting to it is probably a good idea

Comment (ihhub) [reply src/engine/screen.cpp:553]: What's the resolution in fullscreen mode?

Comment (Northfear) [reply src/engine/screen.cpp:553]: The resolution is always 960x544. "Fullscreen" toggle just decides whether the game screen is centered or stretched.

**Comment by zarkone** [comment]
tested on vita, works great for me :+1:

**Review by ihhub — APPROVED** [review]

**Review by oleg-derevenetz — COMMENTED** [review]
LGTM, just one single-click suggestion. BTW, since key mapping changes in this PR may affect Switch, maybe it's worth to update the `docs/README_switch.md` regarding controls. Maybe someone who uses Switch will want to do this as a separate PR.

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by ihhub** [comment]
@Northfear , thank you so much for these changes!

**Pull request merged**